atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClientPool.java

```diff
@@ -204,29 +213,61 @@ public CassandraClientPool(CassandraKeyValueServiceConfig config) {
 
     private CassandraClientPool(CassandraKeyValueServiceConfig config, StartupChecks startupChecks) {
         this.config = config;
-        config.servers().forEach(this::addPool);
         refreshDaemon = Tracers.wrap(PTExecutors.newScheduledThreadPool(1, new ThreadFactoryBuilder()
                 .setDaemon(true)
                 .setNameFormat("CassandraClientPoolRefresh-%d")
                 .build()));
-        refreshDaemon.scheduleWithFixedDelay(() -> {
-            try {
-                refreshPool();
-            } catch (Throwable t) {
-                log.error("Failed to refresh Cassandra KVS pool."
-                        + " Extended periods of being unable to refresh will cause perf degradation.", t);
+        this.startupChecks = startupChecks;
+        isInitialised.compareAndSet(false, false);
+        initialize();
+    }
+
+    @Override
+    public void initialize() {
+        executorService.execute(() -> {
+            while (!isInitialised.get()) {
+                initWithoutThrowing();
             }
-        }, config.poolRefreshIntervalSeconds(), config.poolRefreshIntervalSeconds(), TimeUnit.SECONDS);
+        });
+    }
+
+    private void initWithoutThrowing() {
+        try {
+            config.servers().forEach(this::addPool);
+            refreshDaemon.scheduleWithFixedDelay(() -> {
+                try {
+                    refreshPool();
+                } catch (Throwable t) {
+                    log.error("Failed to refresh Cassandra KVS pool."
+                                    + " Extended periods of being unable to refresh will cause perf degradation.",
+                            t);
+                }
+            }, config.poolRefreshIntervalSeconds(), config.poolRefreshIntervalSeconds(), TimeUnit.SECONDS);
 
-        // for testability, mock/spy are bad at mockability of things called in constructors
-        if (startupChecks == StartupChecks.RUN) {
-            runOneTimeStartupChecks();
+            // for testability, mock/spy are bad at mockability of things called in constructors
+            if (startupChecks == StartupChecks.RUN) {
+                runOneTimeStartupChecks();
+            }
+            refreshPool(); // ensure we've initialized before returning
+            registerAggregateMetrics();
+            isInitialised.compareAndSet(false, true);
+            log.info("Initialization of the ClientPool was successful.");
```

Comment:
why do we even need an compareAndSet here? we could just `set(true)` or make this a `volatile boolean`, since this code should never execute concurrently

(along those lines, we should enforce it never does, probably via the public `initialize` method)